setup.py

```diff
@@ -134,56 +157,47 @@ def load_description():
-        'Intended Audience :: System Administrators',
-        'License :: OSI Approved :: Apache Software License',
-        'Operating System :: MacOS :: MacOS X',
-        'Operating System :: Microsoft :: Windows',
-        'Operating System :: POSIX',
-        'Operating System :: POSIX :: Linux',
-        'Topic :: Internet :: WWW/HTTP',
-        'Topic :: Internet :: WWW/HTTP :: WSGI',
-        # 'Topic :: Internet :: WWW/HTTP :: ASGI',
-        'Topic :: Software Development :: Libraries :: Application Frameworks',
-        'Programming Language :: Python',
-        'Programming Language :: Python :: Implementation :: CPython',
-        'Programming Language :: Python :: Implementation :: PyPy',
-        'Programming Language :: Python :: 3',
-        'Programming Language :: Python :: 3.5',
-        'Programming Language :: Python :: 3.6',
-        'Programming Language :: Python :: 3.7',
-        'Programming Language :: Python :: 3.8',
-        'Programming Language :: Python :: 3.9',
-        'Programming Language :: Cython',
-    ],
-    keywords='wsgi web api framework rest http cloud',
-    author='Kurt Griffiths',
-    author_email='mail@kgriffs.com',
-    url='https://falconframework.org',
-    license='Apache 2.0',
-    packages=find_packages(exclude=['examples', 'tests']),
-    include_package_data=True,
-    zip_safe=False,
-    python_requires='>=3.5',
-    install_requires=REQUIRES,
-    cmdclass=cmdclass,
-    ext_modules=ext_modules,
-    tests_require=['testtools', 'requests', 'pyyaml', 'pytest', 'pytest-runner'],
-    entry_points={
-        'console_scripts': [
-            'falcon-bench = falcon.cmd.bench:main',
-            'falcon-inspect-app = falcon.cmd.inspect_app:main',
-            'falcon-print-routes = falcon.cmd.inspect_app:route_main',
-        ]
-    }
-)
+def run_setup(CYTHON):
+    if CYTHON:
+        cmdclass, ext_modules = get_cython_options()
+    else:
+        cmdclass, ext_modules = {}, []
+
+    setup(
+        long_description=load_description(),
+        cmdclass=cmdclass,
+        ext_modules=ext_modules,
+    )
+
+
+def status_msgs(*msgs):
+    print('*' * 75, *msgs, '*' * 75, sep='\n')
+
+
+if not CYTHON:
+    run_setup(False)
+    if not PYPY:
+        status_msgs('Cython compilation not supported in this environment')
+elif os.environ.get('DISABLE_FALCON_CYTHON'):
+    run_setup(False)
+    status_msgs(
+        'DISABLE_FALCON_CYTHON is set, skipping cython compilation.',
+        'Pure-Python build succeeded.',
+    )
+else:
+    try:
+        run_setup(True)
+    except BuildFailed as exc:
+        status_msgs(
+            exc.__cause__,
+            'Cython compilation could not be completed, speedups are not enabled.',
+            'Failure information, if any, is above.',
+            'Retrying the build without the C extension now.'
+        )
+
+        run_setup(False)
+
+        status_msgs(
+            'Cython compilation could not be completed, speedups are not enabled.',
```

Comment:
Wouldn't this message be repeated twice in the case pure Python build succeeds?

Reply:
Consider that this is only visible if the verbose flag is passed (aka almost no-one)
This is what is printed:
```
    building 'falcon.api_helpers' extension
    ***************************************************************************
    Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
    Cython compilation could not be completed, speedups are not enabled.
    Failure information, if any, is above.
    Retrying the build without the C extension now.
    ***************************************************************************
    running develop
    running egg_info
    writing falcon.egg-info\PKG-INFO
    writing dependency_links to falcon.egg-info\dependency_links.txt
    writing entry points to falcon.egg-info\entry_points.txt
    writing top-level names to falcon.egg-info\top_level.txt
    warning: the 'license_file' option is deprecated, use 'license_files' instead
    adding license file 'LICENSE' (matched pattern 'LICENSE')
    reading manifest file 'falcon.egg-info\SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    writing manifest file 'falcon.egg-info\SOURCES.txt'
    running build_ext
    Creating c:\users\cfede\miniconda3\envs\falcon\lib\site-packages\falcon.egg-link (link to .)
    Adding falcon 3.0.2.dev1 to easy-install.pth file
    Installing falcon-bench-script.py script to C:\Users\cfede\miniconda3\envs\falcon\Scripts
    Installing falcon-bench.exe script to C:\Users\cfede\miniconda3\envs\falcon\Scripts
    Installing falcon-inspect-app-script.py script to C:\Users\cfede\miniconda3\envs\falcon\Scripts
    Installing falcon-inspect-app.exe script to C:\Users\cfede\miniconda3\envs\falcon\Scripts
    Installing falcon-print-routes-script.py script to C:\Users\cfede\miniconda3\envs\falcon\Scripts
    Installing falcon-print-routes.exe script to C:\Users\cfede\miniconda3\envs\falcon\Scripts

    Installed c:\users\cfede\dev\github\falcon
    ***************************************************************************
    Cython compilation could not be completed, speedups are not enabled.
    Pure-Python build succeeded.
    ***************************************************************************
Successfully installed falcon
```

I think in makes sense, maybe we could invert the working:
```
    Pure-Python build succeeded.
    Cython compilation could not be completed, speedups are not enabled.
```